xknx/knxip/srp.py

```diff
@@ -0,0 +1,151 @@
+is satisfied.
+"""
+from __future__ import annotations
+
+from ..exceptions import ConversionError
+from .knxip_enum import DIBServiceFamily, DIBTypeCode, SearchRequestParameterType
+
+
+class Srp:
+    """Search request parameter for a search request extended."""
+
+    SRP_HEADER_SIZE = 2
+    MANDATORY_BIT_INDEX = 0x07
+
+    # service family and service version have 1 byte each
+    SERVICE_PAYLOAD_LENGTH = 2
+    # mac address has 6 bytes
+    MAC_ADDRESS_PAYLOAD_LENGTH = 6
+
+    def __init__(
+        self,
+        srp_type: SearchRequestParameterType,
+        mandatory: bool = True,
+        data: bytes = bytes(),
+    ):
+        """Initialize a SRP."""
+        self.type = srp_type
+        self.mandatory = mandatory
+        self.data = data
+        self.payload_size = Srp.SRP_HEADER_SIZE
+
+        if self.type == SearchRequestParameterType.SELECT_BY_SERVICE:
+            self.validate_payload_length(Srp.SERVICE_PAYLOAD_LENGTH)
+            self.payload_size = Srp.SRP_HEADER_SIZE + Srp.SERVICE_PAYLOAD_LENGTH
+        elif self.type == SearchRequestParameterType.SELECT_BY_MAC_ADDRESS:
+            self.validate_payload_length(Srp.MAC_ADDRESS_PAYLOAD_LENGTH)
+            self.payload_size = Srp.SRP_HEADER_SIZE + Srp.MAC_ADDRESS_PAYLOAD_LENGTH
+        elif self.type == SearchRequestParameterType.REQUEST_DIBS:
+            if not self.data or len(self.data) < 1:
+                raise ConversionError("Srp DIBs are not present.")
+            # If the Client is interested in an odd number of DIBs
+            # it shall add an additional Description Type 0 to make the structure length even
+            if len(self.data) % 2:
+                self.data = self.data + bytes([0])
+            self.payload_size = Srp.SRP_HEADER_SIZE + len(self.data)
+
+    def validate_payload_length(self, size: int) -> None:
+        """Validate the length of the payload."""
+        if not self.data or len(self.data) != size:
+            raise ConversionError(
+                "Srp parameter payload size does not match.",
+                expected_size=size,
+                srp_type=self.type,
+            )
+
+    def __len__(self) -> int:
+        """Get the payload length."""
+        return self.payload_size
+
+    def __bytes__(self) -> bytes:
+        """Convert this SRP to a bytes object."""
+        return (
+            bytes(
+                [
+                    self.payload_size,
+                    (
+                        self.mandatory << Srp.MANDATORY_BIT_INDEX
+                        | self.type.value & Srp.MANDATORY_BIT_INDEX
+                    ),
+                ]
+            )
+            + self.data
+        )
+
+    @staticmethod
+    def from_knx(data: bytes) -> Srp:
+        """Convert the bytes to a SRP object."""
+        if len(data) < Srp.SRP_HEADER_SIZE:
+            raise ConversionError("Data too short for SRP object.")
+
+        size: int = data[0] & 0xFF
```

Comment:
```suggestion
        size: int = data[0]
```
one `bytes` can't be larger than 0xFF